File: Forms/RunTool.vb
Option Strict On
Option Explicit On

Public Class RunTool
    Private Const XML_DATA_FILE As String = "Tools.xml"
    Private ReadOnly _Tools As New Tools()
    Private _ToolsAutocomplete As CustomAutoCompleteCollection

    Private Sub uiButtonOK_Click(ByVal sender As Object, ByVal e As EventArgs) Handles uiButtonOK.Click
        Dim lProcessInfo As New ProcessStartInfo
        Dim lPath As String = _ToolsAutocomplete.GetValue(uiTextBoxCommand.Text, "path")

        If String.IsNullOrEmpty(lPath) Then
            uiErrorProvider.SetError(uiTextBoxCommand, "That QuickTool ID does not exist.")
            uiTextBoxCommand.Text = String.Empty
        End If

        With lProcessInfo
            .FileName = lPath
            .UseShellExecute = True
            .ErrorDialog = True
            .ErrorDialogParentHandle = Handle
        End With

        Try
            Process.Start(lProcessInfo)
            Close()
        Catch ex As Exception
        End Try
    End Sub

    Private Sub uiButtonCancel_Click(ByVal sender As Object, ByVal e As EventArgs) Handles uiButtonCancel.Click
        DialogResult = DialogResult.Cancel
        Close()
    End Sub

    Private Sub uiComboBoxCommand_KeyUp(ByVal sender As Object, ByVal e As KeyEventArgs) Handles uiTextBoxCommand.KeyUp
        Select Case e.KeyCode
            Case Keys.Enter, Keys.Return
                e.Handled = True
                e.SuppressKeyPress = True
                uiButtonOK.PerformClick()
        End Select
    End Sub

    Private Sub QuickRun_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
        _Tools.ReadXml(XMLDataFilePath, XmlReadMode.IgnoreSchema)

        _ToolsAutocomplete = New CustomAutoCompleteCollection(_Tools.QuickRun.CreateDataReader, "shortname")
        uiTextBoxCommand.AutoCompleteCustomSource = _ToolsAutocomplete
    End Sub

    Private Shared Function XMLDataFilePath() As String
        Dim lReturn As String = IO.Path.Combine(My.Application.Info.DirectoryPath, XML_DATA_FILE)

        Return lReturn
    End Function
End Class

File: QuickRun/Forms/RunTool.cs
using QuickRun.Classes;
using QuickRun.DataSources;
using System;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Windows.Forms;
using System.IO;
namespace QuickRun.Forms
{
    public partial class RunTool : Form
    {
        public RunTool()
        {
            InitializeComponent();

            //Initialise event handlers
            uiButtonOK.Click += uiButtonOK_Click;
            uiButtonCancel.Click += uiButtonCancel_Click;
            this.Load += RunTool_Load;
        }

        private const string XML_DATA_FILE = "Tools.xml";
        private readonly Tools _Tools = new Tools();
        private CustomAutoCompleteCollection _ToolsAutocomplete;

        private void uiButtonOK_Click(object sender, EventArgs e)
        {
            ProcessStartInfo lProcessInfo = new ProcessStartInfo();
            string lPath = _ToolsAutocomplete.GetValue(uiTextBoxCommand.Text, "path");

            if (string.IsNullOrEmpty(lPath))
            {
                uiErrorProvider.SetError(uiTextBoxCommand, "That QuickTool ID does not exist.");
                uiTextBoxCommand.Text = string.Empty;
            }

            lProcessInfo.FileName = lPath;
            lProcessInfo.UseShellExecute = true;
            lProcessInfo.ErrorDialog = true;
            lProcessInfo.ErrorDialogParentHandle = Handle;

            try
            {
                Process.Start(lProcessInfo);
                Close();
            }
            catch (Exception ex)
            { }
        }

        private void uiButtonCancel_Click(object sender, EventArgs e)
        {
            DialogResult = DialogResult.Cancel;
            Close();
        }

        private void RunTool_Load(object sender, EventArgs e)
        {
            try
            {
                _Tools.ReadXml(XMLDataFilePath, XmlReadMode.IgnoreSchema);
            }
            catch (Exception ex)
            { }

            _ToolsAutocomplete = new CustomAutoCompleteCollection(_Tools.QuickRun.CreateDataReader(), "shortname");
            uiTextBoxCommand.AutoCompleteCustomSource = _ToolsAutocomplete;
            uiTextBoxCommand.Focus();
        }

        private static string XMLDataFilePath
        {
            get
            {
                string lReturn = Path.Combine(Application.StartupPath, XML_DATA_FILE);

                return lReturn;
            }
        }
    }
}
